File: Stack/implement-queue-using-stacks-232.py
# https://leetcode.com/problems/implement-queue-using-stacks/

# TL;DR:
## This approach uses O(1) for push and amortized O(1) for pop/peek
## The main logic is that we have stack1 for pushing and stack2 for popping
## When stack2 is empty we transfer all elements from stack1 to stack2
## The empty method also needs to check for both stacks to see if the queue is empty

# Complexities:
## Time => O(1) for push, O(1) average - O(n) worst case for pop/peek
## Space => O(n), where n is the amount of numbers passed as input

class MyQueue:

    def __init__(self):
        self.stack1 = []
        self.stack2 = []


    def push(self, x: int) -> None:
        self.stack1.append(x)


    def pop(self) -> int:
        if not self.stack2:
            while self.stack1:
                self.stack2.append(self.stack1.pop())
        return self.stack2.pop()

    def peek(self) -> int:
        if not self.stack2:
            while self.stack1:
                self.stack2.append(self.stack1.pop())
        return self.stack2[-1]

    def empty(self) -> bool:
        return max(len(self.stack1), len(self.stack2)) == 0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: Stack/implement-queue-using-stacks-232.ts
// https://leetcode.com/problems/implement-queue-using-stacks/

// Hint:
// - Separate the push operation from the pop operation in terms of stacks

// TL;DR:
// Use 2 stacks (inStack and outStack) to implement the queue
// Push:
//   - Push the new element to the inStack
// Pop:
//   - If the outStack is empty, transfer all elements from inStack to outStack
//   - Pop the top element from the outStack
// Peek:
//   - If the outStack is empty, transfer all elements from inStack to outStack
//   - Return the top element from the outStack
// Empty:
//   - Return true if both stacks are empty, false otherwise

// Complexities:
// Time => O(1) for push, O(1) amortized for pop and peek, O(1) for empty
// Space => O(n), where n is the number of elements in the stacks

class MyQueue {
	private readonly inStack: number[];
	private readonly outStack: number[];

	constructor() {
		this.inStack = [];
		this.outStack = [];
	}

	push(x: number): void {
		this.inStack.push(x);
	}

	pop(): number {
		if (!this.outStack.length) {
			while (this.inStack.length) {
				this.outStack.push(this.inStack.pop()!);
			}
		}
		return this.outStack.pop()!;
	}

	peek(): number {
		if (!this.outStack.length) {
			while (this.inStack.length) {
				this.outStack.push(this.inStack.pop()!);
			}
		}
		return this.outStack[this.outStack.length - 1];
	}

	empty(): boolean {
		return this.outStack.length === 0 && this.inStack.length === 0;
	}
}
